In [1]:
import cutecharts.charts as cute
import pandas as pd
import numpy as np
import os

#exploring the dataset to see what we can use
books = pd.read_csv("datasets\children_books.csv")

print (f"There are {books['Title'].nunique()} unique books in the {books['Title'].count()} records.")
print("----------------------------")
print(books.info())

There are 41 unique books in the 3269 records.
----------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3269 entries, 0 to 3268
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Title        3269 non-null   object
 1   Author       3096 non-null   object
 2   Desc         3269 non-null   object
 3   Inerest_age  3269 non-null   object
 4   Reading_age  3269 non-null   object
dtypes: object(5)
memory usage: 127.8+ KB
None


It seems like there are only 41 unique books in the children books dataset despite having 3269 rows of data with some Author data missing.
Let's explore if the different unique books have have different meaningful data on the other columns. 

In [2]:
#I will use a few unique title to explore further
books['Title'].unique()

array(['The Girl Who Drank the Moon', 'Time Between Us',
       'Girl Out of Water', 'Captive', 'The School of Music',
       'The Crystal Stair', 'Dark Inside', 'Kite Spirit',
       'I Capture the Castle', 'Butter', 'Dead Jealous',
       'A Flash of Blue', 'Jet Black Heart', "The Sin Eater's Daughter",
       'King Lear - Manga Shakespeare', 'Frankenstein',
       'The Song from Somewhere Else', "Amy & Roger's Epic Detour",
       'The Changes Trilogy', 'The Train to Impossible Places',
       'Half Brother', 'Holding Up the Universe', 'Pop-Up Frankenstein',
       'Mister Creecher', 'The Complete Calvin and Hobbes',
       "Don't Pick on Me!", 'Five Hundred Miles', 'Rules for a Knight',
       'Paper Avalanche', 'Ghost World', 'Invictus',
       'The Name of the Star', 'Spy for the Queen of Scots',
       'Jimmy Corrigan:', 'The Martian Girl', 'Buffalo Soldier',
       'Witchstruck', 'Flyaway', 'Etiquette and Espionage',
       'The Sky is Everywhere', 'Glass Sword'], dtype=object)

In [3]:
example_one = books[books['Title']=='Time Between Us']
example_one

Title                Author  \
1     Time Between Us  Tamara Ireland Stone   
38    Time Between Us  Tamara Ireland Stone   
75    Time Between Us  Tamara Ireland Stone   
112   Time Between Us  Tamara Ireland Stone   
158   Time Between Us  Tamara Ireland Stone   
...               ...                   ...   
3088  Time Between Us  Tamara Ireland Stone   
3123  Time Between Us  Tamara Ireland Stone   
3160  Time Between Us  Tamara Ireland Stone   
3197  Time Between Us  Tamara Ireland Stone   
3233  Time Between Us  Tamara Ireland Stone   

                                                   Desc Inerest_age  \
1     Sixteen-year-old Anna is struggling to underst...        13+    
38    Sixteen-year-old Anna is struggling to underst...        13+    
75    Sixteen-year-old Anna is struggling to underst...        13+    
112   Sixteen-year-old Anna is struggling to underst...        13+    
158   Sixteen-year-old Anna is struggling to underst...        13+    
...                                                 ...         ...   
3088  Sixteen-year-old Anna is struggling to underst...        13+    
3123  Sixteen-year-old Anna is struggling to underst...        13+    
3160  Sixteen-year-old Anna is struggling to underst...        13+    
3197  Sixteen-year-old Anna is struggling to underst...        13+    
3233  Sixteen-year-old Anna is struggling to underst...        13+    

                             Reading_age  
1     12+                                 
38    12+                                 
75    12+                                 
112   12+                                 
158   12+                                 
...                                  ...  
3088  12+                                 
3123  12+                                 
3160  12+                                 
3197  12+                                 
3233  12+                                 

[88 rows x 5 columns]

In [4]:
example_two = books[books['Title']=='Butter']
example_two

Title           Author  \
9     Butter  Erin Jade Lange   
46    Butter  Erin Jade Lange   
83    Butter  Erin Jade Lange   
120   Butter  Erin Jade Lange   
166   Butter  Erin Jade Lange   
...      ...              ...   
3095  Butter  Erin Jade Lange   
3131  Butter  Erin Jade Lange   
3168  Butter  Erin Jade Lange   
3205  Butter  Erin Jade Lange   
3241  Butter  Erin Jade Lange   

                                                   Desc Inerest_age  \
9     Isolated and bullied at school, with a dysfunc...        13+    
46    Isolated and bullied at school, with a dysfunc...        13+    
83    Isolated and bullied at school, with a dysfunc...        13+    
120   Isolated and bullied at school, with a dysfunc...        13+    
166   Isolated and bullied at school, with a dysfunc...        13+    
...                                                 ...         ...   
3095  Isolated and bullied at school, with a dysfunc...        13+    
3131  Isolated and bullied at school, with a dysfunc...        13+    
3168  Isolated and bullied at school, with a dysfunc...        13+    
3205  Isolated and bullied at school, with a dysfunc...        13+    
3241  Isolated and bullied at school, with a dysfunc...        13+    

                             Reading_age  
9     13+                                 
46    13+                                 
83    13+                                 
120   13+                                 
166   13+                                 
...                                  ...  
3095  13+                                 
3131  13+                                 
3168  13+                                 
3205  13+                                 
3241  13+                                 

[88 rows x 5 columns]

In [5]:
example_three = books[books['Title']=='The Train to Impossible Places']
example_three

Title  \
19    The Train to Impossible Places   
56    The Train to Impossible Places   
93    The Train to Impossible Places   
130   The Train to Impossible Places   
176   The Train to Impossible Places   
...                              ...   
3105  The Train to Impossible Places   
3141  The Train to Impossible Places   
3178  The Train to Impossible Places   
3215  The Train to Impossible Places   
3251  The Train to Impossible Places   

                                        Author  \
19    P.G. Bell Illustrator: Flavia Sorrentino   
56    P.G. Bell Illustrator: Flavia Sorrentino   
93    P.G. Bell Illustrator: Flavia Sorrentino   
130   P.G. Bell Illustrator: Flavia Sorrentino   
176   P.G. Bell Illustrator: Flavia Sorrentino   
...                                        ...   
3105  P.G. Bell Illustrator: Flavia Sorrentino   
3141  P.G. Bell Illustrator: Flavia Sorrentino   
3178  P.G. Bell Illustrator: Flavia Sorrentino   
3215  P.G. Bell Illustrator: Flavia Sorrentino   
3251  P.G. Bell Illustrator: Flavia Sorrentino   

                                                   Desc Inerest_age  \
19    Suzy Smith is very surprised when a huge magic...      10-12    
56    Suzy Smith is very surprised when a huge magic...      10-12    
93    Suzy Smith is very surprised when a huge magic...      10-12    
130   Suzy Smith is very surprised when a huge magic...      10-12    
176   Suzy Smith is very surprised when a huge magic...      10-12    
...                                                 ...         ...   
3105  Suzy Smith is very surprised when a huge magic...      10-12    
3141  Suzy Smith is very surprised when a huge magic...      10-12    
3178  Suzy Smith is very surprised when a huge magic...      10-12    
3215  Suzy Smith is very surprised when a huge magic...      10-12    
3251  Suzy Smith is very surprised when a huge magic...      10-12    

                               Reading_age  
19    10-12                                 
56    10-12                                 
93    10-12                                 
130   10-12                                 
176   10-12                                 
...                                    ...  
3105  10-12                                 
3141  10-12                                 
3178  10-12                                 
3215  10-12                                 
3251  10-12                                 

[85 rows x 5 columns]

Upon exploring 3 examples, the titles are repeated with all the the other columns having no unique value.
Thus having multiple repeated rows is not meaningful.
I will now proceed to keep only those that are unique.

In [6]:
#to drop the books with no Author data
books.drop_duplicates(subset=None, keep='first', inplace=True, ignore_index=False)

books

Title  \
0       The Girl Who Drank the Moon   
1                   Time Between Us   
2                 Girl Out of Water   
3                           Captive   
4               The School of Music   
5                 The Crystal Stair   
6                       Dark Inside   
7                       Kite Spirit   
8              I Capture the Castle   
9                            Butter   
10                     Dead Jealous   
11                  A Flash of Blue   
12                  Jet Black Heart   
13         The Sin Eater's Daughter   
14    King Lear - Manga Shakespeare   
15                     Frankenstein   
16     The Song from Somewhere Else   
17        Amy & Roger's Epic Detour   
18              The Changes Trilogy   
19   The Train to Impossible Places   
20                     Half Brother   
21          Holding Up the Universe   
22              Pop-Up Frankenstein   
23                  Mister Creecher   
24   The Complete Calvin and Hobbes   
25                Don't Pick on Me!   
26               Five Hundred Miles   
27               Rules for a Knight   
28                  Paper Avalanche   
29                      Ghost World   
30                         Invictus   
31             The Name of the Star   
32       Spy for the Queen of Scots   
33                  Jimmy Corrigan:   
34                 The Martian Girl   
35                  Buffalo Soldier   
36                      Witchstruck   
148                         Flyaway   
149         Etiquette and Espionage   
150           The Sky is Everywhere   
151                     Glass Sword   

                                                Author  \
0                                       Kelly Barnhill   
1                                 Tamara Ireland Stone   
2                                        Nat Luurtsema   
3                                         A J Grainger   
4    Rachel Bowen and Meurig Bowen Illustrator: Dan...   
5                                     Catherine Fisher   
6                                         Jeyn Roberts   
7                                     Sita Brahmachari   
8                                          Dodie Smith   
9                                      Erin Jade Lange   
10                                        Sharon Jones   
11                                        Maria Farrer   
12                                       Teresa Flavin   
13                                   Melinda Salisbury   
14                                 William Shakespeare   
15                                        Mary Shelley   
16               AF Harrold\nIllustrator: Levi Pinfold   
17                                       Morgan Matson   
18                                     Peter Dickinson   
19            P.G. Bell Illustrator: Flavia Sorrentino   
20                                       Kenneth Oppel   
21                                      Jennifer Niven   
22   Claire Hawcock (adapted from Mary Shelley) Ill...   
23                                     Chris Priestley   
24                                                 NaN   
25                                     Rosemary Stones   
26                                        Kevin Brooks   
27                 Ethan Hawke Illustrator: Ryan Hawke   
28                                     Lisa Williamson   
29                                       Daniel Clowes   
30                                        Ryan Graudin   
31                                     Maureen Johnson   
32                                     Theresa Breslin   
33                                          Chris Ware   
34                                                 NaN   
35                                       Tanya Landman   
36                                       Victoria Lamb   
148                                   Lucy Christopher   
149                                      Gail Carriger   
150                                       Jandy Nelson   
151          

In [7]:
books.dropna(axis=0, how='any', thresh=None, subset=None, inplace=True)

books.reset_index(drop=True, inplace=True)
books

Title  \
0      The Girl Who Drank the Moon   
1                  Time Between Us   
2                Girl Out of Water   
3                          Captive   
4              The School of Music   
5                The Crystal Stair   
6                      Dark Inside   
7                      Kite Spirit   
8             I Capture the Castle   
9                           Butter   
10                    Dead Jealous   
11                 A Flash of Blue   
12                 Jet Black Heart   
13        The Sin Eater's Daughter   
14   King Lear - Manga Shakespeare   
15                    Frankenstein   
16    The Song from Somewhere Else   
17       Amy & Roger's Epic Detour   
18             The Changes Trilogy   
19  The Train to Impossible Places   
20                    Half Brother   
21         Holding Up the Universe   
22             Pop-Up Frankenstein   
23                 Mister Creecher   
24               Don't Pick on Me!   
25              Five Hundred Miles   
26              Rules for a Knight   
27                 Paper Avalanche   
28                     Ghost World   
29                        Invictus   
30            The Name of the Star   
31      Spy for the Queen of Scots   
32                 Jimmy Corrigan:   
33                 Buffalo Soldier   
34                     Witchstruck   
35                         Flyaway   
36         Etiquette and Espionage   
37           The Sky is Everywhere   
38                     Glass Sword   

                                               Author  \
0                                      Kelly Barnhill   
1                                Tamara Ireland Stone   
2                                       Nat Luurtsema   
3                                        A J Grainger   
4   Rachel Bowen and Meurig Bowen Illustrator: Dan...   
5                                    Catherine Fisher   
6                                        Jeyn Roberts   
7                                    Sita Brahmachari   
8                                         Dodie Smith   
9                                     Erin Jade Lange   
10                                       Sharon Jones   
11                                       Maria Farrer   
12                                      Teresa Flavin   
13                                  Melinda Salisbury   
14                                William Shakespeare   
15                                       Mary Shelley   
16              AF Harrold\nIllustrator: Levi Pinfold   
17                                      Morgan Matson   
18                                    Peter Dickinson   
19           P.G. Bell Illustrator: Flavia Sorrentino   
20                                      Kenneth Oppel   
21                                     Jennifer Niven   
22  Claire Hawcock (adapted from Mary Shelley) Ill...   
23                                    Chris Priestley   
24                                    Rosemary Stones   
25                                       Kevin Brooks   
26                Ethan Hawke Illustrator: Ryan Hawke   
27                                    Lisa Williamson   
28                                      Daniel Clowes   
29                                       Ryan Graudin   
30                                    Maureen Johnson   
31                                    Theresa Breslin   
32                                         Chris Ware   
33                                      Tanya Landman   
34                                      Victoria Lamb   
35                                   Lucy Christopher   
36                                      Gail Carriger   
37                                       Jandy Nelson   
38                                   Victoria Aveyard   

                                                 Desc Inerest_age  \
0   Every year, the evil Protectorate offers a bab...      10-14    
1   Sixteen-year-old Anna is struggling to underst...        13+    
2   Lou Brown's swim

This means I only have 39 Book titles to work with. I will set this aside and explore the other dataset first.

In [8]:
stories = pd.read_csv("datasets\children_stories.csv", encoding = "ISO-8859-1")

print (f"There are {stories['names'].nunique()} unique book names in the {stories['names'].count()} records.")
print("----------------------------")
print(stories.info())

There are 390 unique book names in the 430 records.
----------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 430 entries, 0 to 429
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   names   430 non-null    object
 1   cats    430 non-null    object
 2   desc    430 non-null    object
dtypes: object(3)
memory usage: 10.2+ KB
None


There are more to work with on the Children stories dataset.
Again, there are 390 unique values out of 430 records. 
Let's see if there are meaningful or duplicated data.

In [9]:
duplicated_stories = stories[stories['names'].duplicated()]
duplicated_stories.sort_values(by=['names'])


names          cats  \
211                                 Alfie Gives a Hand        Age 3+   
152                                     Alfie Outdoors        Age 3+   
207                                    Baby's Bathtime        Age 0+   
197                       Do You Want To Be My Friend?        Age 3+   
187                           Elmer and the Lost Teddy        Age 0+   
195                                    Elmer's Friends        Age 0+   
149                                   Father Christmas        Age 2+   
190                                  From Acorn to Zoo        Age 2+   
169                                Horton Hears a Who!        Age 3+   
130                                                Hug        Age 0+   
192                                     I Like It When       Age 1-4   
156                                   I Want My Potty!        Age 2+   
178                                Incey Wincey Spider        Age 1+   
139                                             Kipper       Age 2-4   
143                                  Kipper's Birthday        Age 2+   
206                                           Macavity        Age 3+   
151                            Melric and the Sorcerer        Age 4+   
172             Melric the Magician Who Lost His Magic        Age 3+   
157                                   Mockingbird Song        Age 0+   
146                            Mrs Wobble the Waitress       Age 3-6   
133                                My Dad is Brilliant        Age 1+   
65                                      Nursery Rhymes       Age 0-4   
147                                     Nursery Rhymes        Age 0+   
159                              Oi! Get Off Our Train        Age 3+   
182                                         Owl Babies       Age 3-7   
191                                        Pat-a-Cake!        Age 0+   
202                               Princess Smartypants        Age 3+   
153                                Quentin Blake's ABC        Age 2+   
138                                         Snow Bears        Age 2+   
163                           The Emperors New Clothes        Age 2+   
148  The Further Adventures of the Owl and the Puss...        Age 3+   
298                                       The Iron Man      Age 9-11   
216                                        The Snowman      Age 7-10   
194                                        The Snowman        Age 3+   
193                 The Usborne Book of Nursery Rhymes  Age 6months+   
136                         We're Going on a Bear Hunt        Age 2+   
181                         We're Going on a Lion Hunt        Age 3+   
179                          What's the Time, Mr Wolf?        Age 0+   
180                                             Window        Age 4+   
186                               Za-za's Baby Brother        Age 2+   

                                                  desc  
211  Shirley Hughes' much-loved Alfie goes to his f...  
152  One sunny day Alfie decides to explore the bot...  
207  Bold, colourful and great fun, this simple and...  
197  'Do you want to be my friend?' Asked just once...  
187  When Baby Elephant loses his teddy and can't s...  
195  A colourful board book in which the ever-popul...  
149  It's December 24th. But for poor Father Christ...  
190  This is an alphabet book with a difference, in...  
169  Now a major film, Horton Hears a Who! is a hil...  
130  Bobo the chimp wanders sadly around the jungle...  
192  The pleasure of parent-child activities like s...  
156  The Little Princess hates nappies, and thinks ...  
178  Part of the Sign and Singalong series of brigh...  
139  One day, Kipper the dog finds that he just can...  
143  Kipper is working hard preparing for his birth...  
206  Escaped from Old Possum's Book of Practical Ca...  
151  The king wants his magician, Melric, to make h...  
172  Melric, the King’s magician, uses his magic ...  
157  Celebrat

From the above, its observed that some stories are duplicated with the similar cats and same desc values, with a few just with similar stories name such as "Nursery Rhymes" and "The Snowman".
Before removing the duplicated rows, I would like to explore further on the cats column.

In [10]:
print(f"There are {stories['cats'].nunique()} unique categories")
print("----------------------------")
print(stories['cats'].unique())

There are 59 unique categories
----------------------------
['Age 2-9' 'Age 1-2' 'Age 1-3' 'Age 0-2' 'Age  0-3' 'Age 0-3' 'Age 3-5'
 'Age 0+' 'Age 1-5' 'Age 1-4' 'Age 2-6' 'Age 2-7' 'Age 3-6' 'Age 0-4'
 'Age 2-5' 'Age 4-6' 'Age 0-5' 'Age 2+' 'Age 4+' 'Age 3+' 'Age 3-7'
 'Age 3-4' 'Age 1-6' 'Age 0-6' 'Age 4-7' 'Age 5-8' 'Age 1+' 'Age 5+'
 'Age 4-11' 'Age 2-4' 'Age 4-5' 'Age 4-8' 'Age 6months+' 'Age 6-8'
 'Age 6-11' 'Age 8-12' 'Age 7-10' 'Age 7-12' 'Age 7+' 'Age 5-9' 'Age 6+'
 'Age 7-9' 'Age 8+' 'Age 7-11' 'Age 6-9' 'Age 9-14' 'Age 9-11' 'Age 9+'
 'Age 10+' 'Age 11+' 'Age 13+' 'Age 12+' 'Age ' 'Age 9-13' 'Age 11-14'
 'Age 10-14' 'Age 9-12' 'Age 8-11' 'Age 11-15']


Seeing how there are 59 unique categories, the cats column needs to be cleaned up and categorized accordingly.
I have deviced to go with the guide on https://www.healthychildren.org/English/ages-stages/Pages/default.aspx
where I would then categorize the age groups according to the appropriate stage as follows:
Baby: 0-1 year
Toddler: 1-3 years
Preschool: 3-5 years
Gradeschooler: 5-12 years
Teen: 12-18 years

As there are some age gaps in between, I will take account of the start age and merge Baby and Toddler category into just Toddler.
My Stages column would look like:
Toddler: 0-3 years
Preschool: 3-5 years
Gradeschooler: 5-12 years
Teen: 12-18 years

In [11]:
groups = {'toddler' : ['Age 1-2','Age 1-3','Age 0-2','Age  0-3','Age 0-3','Age 0+', 
              'Age 0-4', 'Age 0-5', 'Age 2+', 'Age 3+','Age 1+','Age 6months+'],
'preschool': ['Age 3-5','Age 1-5','Age 1-4','Age 2-6','Age 3-6','Age 2-5',
                'Age 4-6','Age 4+','Age 3-4','Age 1-6','Age 0-6','Age 2-4','Age 4-5'],
'gradeschool' : ['Age 2-9','Age 2-7','Age 3-7','Age 4-7','Age 5-8','Age 5+', 'Age 4-11'  
                    'Age 4-8','Age 6-8','Age 6-11','Age 8-12','Age 7-10','Age 7-12','Age 7+',
                    'Age 5-9','Age 6+','Age 7-9','Age 8+','Age 7-11','Age 6-9','Age 9-14',
                    'Age 9-11','Age 9+','Age 10+','Age 11+','Age 9-12','Age 8-11'],
'teens': ['Age 13+','Age 12+','Age 9-13','Age 11-14','Age 10-14','Age 11-15'],
}
stories["stage"] = (stories["cats"]
    .apply(lambda x: [key for key in groups.keys() if x in groups[key]])
    .str[0]
    .fillna("allages")
)

In [12]:
stories['stage'].value_counts()

gradeschool    227
toddler        105
preschool       72
teens           20
allages          6
Name: stage, dtype: int64

Now that we have the ages categorized according to stages, I return to look at the duplicated stories. I want to remove all the duplicated stories with the same description

In [13]:
stories.drop_duplicates(subset='desc', inplace=True)

In [14]:
#running this again to see what is left that is duplicated
duplicated_stories = stories[stories['desc'].duplicated()]
duplicated_stories.count()

names    0
cats     0
desc     0
stage    0
dtype: int64

After performing the above, only stories with the same name but different description remains.

In [26]:
#see total rows in the cleaned dataframe
len(stories.index)

393